[Material][Menu] Add Nested Menu demo

sai6855 opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

preview: https://deploy-preview-37570--material-ui.netlify.app/material-ui/react-menu#nested-menu

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

Related to #11723

mnajdova commented:
This is an interesting PR, it's great to experiment with this feature in a demo first. A couple of thoughts:
1. Should we add this demo in Base UI too? Once we have the support internally, we can update it
2. I would create a wrapper for the `MenuItem` component inside the demo so that the code looks cleaner - this is what typically developers would do in their code-base
3. Link the aria spec we are following
4. Add some end-to-end tests on the demo to ensure we are not going to break it

sai6855 commented:
>Should we add this demo in Base UI too? Once we have the support internally, we can update it

Yes, plan was to first finalize functionality and approach, once these are finalized we can add to both joy and base ui.

>I would create a wrapper for the MenuItem component inside the demo so that the code looks cleaner - this is what typically developers would do in their code-base

At first i tried with `Menu` and `MenuItem`, but couldn't match the functionality with standard `Nested Menus`. I'll explain what was the issue once PR is ready.

>Link the aria spec we are following
Tbh, i'm not following any specifc aria-spec, just trying to replicate the functionality of Nested Menus of other popular libraries and chrome's and mac's Menu.

>Add some end-to-end tests on the demo to ensure we are not going to break it
sure.

sai6855 commented:
@mnajdova PR is ready for review, but i'm not sure why CI is failing [here](https://app.circleci.com/pipelines/github/mui/material-ui/100814/workflows/d539dcea-1b29-4b08-afb8-2f2cab7eb24b/jobs/537113?invite=true#step-110-1044). can you help me to fix it

michaldudak commented:
Nice! I guess we could write that we plan to implement this in the library at some point, but until then, developers can create their own using the code below.

Few things I've noticed:

* Hovering over the expanded menu triggers popups from the demo toolbar:
  <img width="230" alt="image" src="https://github.com/mui/material-ui/assets/4696105/0a926465-e54c-44b4-977d-e5c370691aaa">
* The menu can't be closed with an <kbd>Esc</kbd> key
* Screen readers do not announce that an item has a nested submenu
* Clicking on an item that has a submenu closes the whole menu. Native OS menus don't work like that

sai6855 commented:
@michaldudak resolved all your suggestions.

can you try below thing again, i think it got accidently fixed. i don't see tooltip on hover now
>Hovering over the expanded menu triggers popups from the demo toolbar:

https://github.com/mui/material-ui/assets/60743144/b90f4f0e-62d8-4dc1-bdf9-74c3ee9e753b

michaldudak commented:
Looks better, but I still can't close it with an <kbd>Esc</kbd> key.
Also, we're missing the `aria-expanded` attribute on items with visible submenus (see https://www.w3.org/WAI/ARIA/apg/patterns/menubar/#wai-ariaroles,states,andproperties).
We could also drop the `aria-haspopup="false"` on elements without a submenu.

oliviertassinari commented:
if we could handle https://twitter.com/diegohaz/status/1283558204178407427, it would make a huge difference. For example, Notion doesn't have this, their nested menu is a **horrible** to use, I waste so much time on this behavior.

sai6855 commented:
> if we could handle https://twitter.com/diegohaz/status/1283558204178407427, it would make a big difference. Each time this is not implemented in a product, you can notice instantly they didn't do it, it make the experience soo frustrating.

@oliviertassinari  i think i made it working as described in this tweet, although i took different approach, but it is working


https://github.com/mui/material-ui/assets/60743144/6a89d9d4-bcfe-4003-8bb5-62f84812683c


@michaldudak addressed all your suggestions.

michaldudak commented on docs/data/material/components/menus/NestedMenu.tsx:
```diff
@@ -0,0 +1,287 @@
+  Typography,
+} from '@mui/material';
+
+export default function BasicMenu() {
```
The exported function name should be the same as the file name.
There is a linter rule for this on master. Make sure to merge in the latest one.

sai6855 replied:
done

michaldudak commented on docs/data/material/components/menus/NestedMenu.tsx:
```diff
@@ -0,0 +1,287 @@
+} from '@mui/material';
+
+export default function BasicMenu() {
+  const MENU_LEVELS = 3;
```
It would be great to have this in a form of a reusable component, so it's easy to copy and paste from our docs. So have a `<NestedMenu>` with all the logic that accepts the `options` prop.

michaldudak reviewed (commented):
The submenus open (and sometimes close) with a long delay - this looks pretty bad UX-wise. It used to work much better, so likely there's a regression.

sai6855 commented:
> The submenus open (and sometimes close) with a long delay - this looks pretty bad UX-wise. It used to work much better, so likely there's a regression.

Ah okay, delay was added to add this funcationality (https://twitter.com/diegohaz/status/1283558204178407427) but looks like UX has gotten bad. i've changed delay from 50ms to 20ms, can you try it now?

ZeeshanTamboli reviewed (commented):
I can still see the popup from the demo toolbar on Windows OS. But it is showing in all the demos.

https://github.com/mui/material-ui/assets/20900032/d4b269b2-0a1a-4121-b040-6f1a97fd8711

sai6855 commented:
> I can still see the popup from the demo toolbar on Windows OS. But it is showing in all the demos.
> 
>  NestedMenu.PR.mp4

Can you test this in mui.com? I don't think this PR introduced this bug

ZeeshanTamboli commented:
It's weird that I can see it in https://deploy-preview-37570--material-ui.netlify.app/material-ui/react-menu/#nested-menu, but I can't see it locally on your branch - http://localhost:3000/material-ui/react-menu/#nested-menu.

sai6855 commented:
> It's weird that I can see it in https://deploy-preview-37570--material-ui.netlify.app/material-ui/react-menu/#nested-menu, but I can't see it locally on your branch - http://localhost:3000/material-ui/react-menu/#nested-menu. Also, I can't see on mui.com.

@ZeeshanTamboli  i can see same behaviour in all preview url's, below video is from this PR (https://github.com/mui/material-ui/pull/38123) and you can see same behaviour on hover

https://github.com/mui/material-ui/assets/60743144/9115dc4c-1fa5-4ef2-8240-a08436687c18

michaldudak commented on docs/data/material/components/menus/NestedMenu.tsx:
```diff
@@ -0,0 +1,297 @@
+                                )
+                              : undefined
+                          }
+                          onClick={() => {
```
It would be more readable if the event handlers were not defined inline.

michaldudak commented on test/e2e/index.test.ts:
```diff
@@ -239,6 +239,29 @@ describe('e2e', () => {
+      await screen.getByText('Lemonade');
+    });
+  });
   describe('<TextareaAutosize />', () => {
```
```suggestion

  describe('<TextareaAutosize />', () => {
```

michaldudak commented on docs/data/material/components/menus/NestedMenu.tsx:
```diff
@@ -0,0 +1,297 @@
+    </React.Fragment>
+  );
+}
+export default function NestedMenu() {
```
```suggestion

export default function NestedMenu() {
```

michaldudak commented on docs/data/material/components/menus/NestedMenu.tsx:
```diff
@@ -0,0 +1,297 @@
+      ),
+    }));
+  };
+  const handleClose = (level: number) => {
```
```suggestion

  const handleClose = (level: number) => {
```

michaldudak commented on docs/data/material/components/menus/NestedMenu.js:
```diff
@@ -0,0 +1,293 @@
+}
+
+SubMenu.propTypes = {
+  MENU_LEVELS: PropTypes.number.isRequired,
```
Why is this uppercase? Props are generally constant anyway.

michaldudak reviewed (commented):
The behavior generally looks good, although I also can see the popup problem. Also, I don't think we have what Olivier suggested in https://github.com/mui/material-ui/pull/37570#issuecomment-1615849903, but we could refine it in another PR.

I still have a couple of remarks.

oliviertassinari commented:
> I don't think we have what Olivier suggested in https://github.com/mui/material-ui/pull/37570#issuecomment-1615849903

1. It doesn't, I can't select "Alcoholic", or any value below, this is not usable in production:

https://github.com/mui/material-ui/assets/3165635/27cbc720-6b18-4cc5-9eff-4e4804f9f814

https://deploy-preview-37570--material-ui.netlify.app/material-ui/react-menu/#nested-menu
Our exigent users will notice this miles away. I think it's better to point users to use another solution in the current state, so I think this should be part of the MVP of the component, in the first iteration.

2. Why invest energy in this for Material UI? Developers have been asking about this for 5 years #11723. I think they can wait one more year. I think it would make more sense to focus on bringing this to Base UI, assuming that we can use its menu with Material UI v6. We need to fill the gap with https://www.radix-ui.com/primitives/docs/components/dropdown-menu.

3. The icons feel too large, and it feels that it does not have enough margin left. Can we try with a few tweaks?

<img width="151" alt="Screenshot 2023-08-31 at 00 08 27" src="https://github.com/mui/material-ui/assets/3165635/0cd710b7-bbdf-4b19-aa6b-c5c13fc0bd55">
